README.md
[![arduino-library-badge](https://www.ardu-badge.com/badge/Namino_Library.svg?)](https://www.ardu-badge.com/Namino_Library)

# Namino_Library

Namino, getting started

The Namino boards can be programmed in Windows / Linux systems, using the Arduino IDE and VS code (PlatformIO) IDE environments. These have their own installation guides, already established. The aim of the guide is how to integrate existing tutorials with Namino features.
Presentation

Our Namino system has an ESP32-S3 integrated to front end/ terminal block capable of connecting, to a machine, using 24V I/O industrial signals, 0-10V, 0-20mA, thermocouples, RTD, PTC, NTC, load cell. The front end is managed by an NXP microcontroller, with its own logic, connected to the SPI interface. Furthermore, a connector for the 3.3V Arduino shields is available.

docs/changelog.rst
Change log
==========

Added
~~~~~~~

* First public release

1.0.18_ |--| 2023-07-14
-----------------------

.. |--| unicode:: U+2013 .. EN DASH
    
.. _1.0.18: https://github.com/namino-cc/Namino_Library/compare

src/namino_arancio.cpp
/*
MIT License

Copyright (c) 2023 Namino Team, version: 1.0.18 @ 2023-07-07

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

#include <Arduino.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>
#include <unistd.h>
#include <string.h>
#include <namino_arancio.h>

#ifndef NAMINO_ARANCIO_BOARD
#define NAMINO_ARANCIO_BOARD
#endif

#if CONFIG_IDF_TARGET_ESP32S2 || CONFIG_IDF_TARGET_ESP32S3
#define VSPI FSPI
#endif

namino_arancio::namino_arancio() {

}

void namino_arancio::begin(uint32_t clock, uint8_t miso, uint8_t mosi, uint8_t sck, uint8_t ss) {
    _miso = miso;
    _mosi = mosi;
    _sck = sck;
    _ss = ss;

    pinMode(_ss, OUTPUT);
    digitalWrite(_ss, HIGH);

    spiSettings = SPISettings(clock, MSBFIRST, SPI_MODE0);
    vspi = new SPIClass(VSPI);
    vspi->begin(_sck, _miso, _mosi, _ss);
    reg[WR_COUNTER] = 0;
}

void namino_arancio::begin(uint32_t clock) {
    _miso = MISO;
    _mosi = MOSI;
    _sck = SCK;
    _ss = SS;

    pinMode(_ss, OUTPUT);
    digitalWrite(_ss, HIGH);

    spiSettings = SPISettings(clock, MSBFIRST, SPI_MODE0);
    vspi = new SPIClass(VSPI);
    vspi->begin(_sck, _miso, _mosi, _ss);
    reg[WR_COUNTER] = 0;
}

void namino_arancio::end() {
    vspi->end();
}

float_t namino_arancio::readAnalogIn(int channel) {
    uint16_t value = 0;
    uint8_t conf = 0;

    switch(channel) {
        case 1:
            value = reg[RO_ANALOG_IN_CH01];
            conf = reg[WR_ANALOG_IN_CH01_CONF];
            switch(conf) {
                case CH01_DISABLED:
                  return 0;
                break;
                case CH01_CURRENT:
                  return value / 1000.0; // place here um conversion
                break;
                case CH01_VOLTAGE:
                  return value / 1000.0; // place here um conversion
                break;
                default:
                  return 0;
                break;
            }
        break;
        case 2:
            value = reg[RO_ANALOG_IN_CH02];
            conf = reg[WR_ANALOG_IN_CH02_CONF];
            switch(conf) {
                case CH02_DISABLED:
                  return 0;
                break;
                case CH02_CURRENT:
                  return value / 1000.0; // place here um conversion
                break;
                case CH02_VOLTAGE:
                  return value / 1000.0; // place here um conversion
                break;
                default:
                  return 0;
                break;
            }
        break;
        case 3:
            value = reg[RO_ANALOG_IN_CH03];
            conf = reg[WR_ANALOG_IN_CH03_CONF];
            switch(conf) {
                case CH03_DISABLED:
                  return 0;
                break;
                case CH03_CURRENT:
                  return value / 1000.0; // place here um conversion
                break;
                case CH03_VOLTAGE:
                  return value / 1000.0; // place here um conversion
                break;
                default:
                  return 0;
                break;
            }
        break;
        case 4:
            value = reg[RO_ANALOG_IN_CH04];
            conf = reg[WR_ANALOG_IN_CH04_CONF];
            switch(conf) {
                case CH04_DISABLED:
                  return 0;
                break;
                case CH04_CURRENT:
                  return value / 1000.0; // place here um conversion
                break;
                case CH04_VOLTAGE:
                  return value / 1000.0; // place here um conversion
                break;
                default:
                  return 0;
                break;
            }
        break;
        case 5:
            value = reg[RO_ANALOG_IN_CH05];
             // place here um conversion
            return value / 10.0;
        break;
        default:
            return 0;
    };
    return 0;
}

int32_t namino_arancio::readEncoder() {
    return (reg[RO_ENCODER_COUNTER_H] << 16) | reg[RO_ENCODER_COUNTER_L];
}

uint32_t namino_arancio::readLifeTime() {
    return (reg[RO_LIFE_TIME_H] << 16) | reg[RO_LIFE_TIME_L];
}

uint16_t namino_arancio::fwVersion() {
  return (reg[RO_NAMINO_ID]);
}

uint16_t namino_arancio::boardType() {
  return (reg[RO_ANALOG_FRONT_END_ID]);
}

bool namino_arancio::isReady() {
    return reg[RO_ANALOG_FRONT_END_ID] == 0x2904;
}

void namino_arancio::resetEncoderCounter() {
    reg[WR_ENCODER_SET_L] = 0;
    reg[WR_ENCODER_SET_H] = 0;
    reg[WR_CONFIG_REGISTER] = 1;
}

void namino_arancio::setEncoderCounter(int32_t value) {
    reg[WR_ENCODER_SET_L] = value & 0x0000FFFF;
    reg[WR_ENCODER_SET_H] = (value & 0xFFFF0000) >> 16;
    reg[WR_CONFIG_REGISTER] = 1;
}

bool namino_arancio::readDigIn(int channel) {
    if (channel >= 1 && channel <= 8) {
        channel--;
        return (reg[RO_DIG_IN_1_16] & (1 << channel)) >> channel;
    }
    return false;
}

bool namino_arancio::readDigOut(int channel) {
    if (channel >= 1 && channel <= 8) {
        channel--;
        return (reg[WR_DIG_OUT_1_16] & (1 << channel)) >> channel;
    }
    return false;
}

bool namino_arancio::readEncoderSwitch() {
    return reg[RO_ENCODER_SWITCH] == 1;
}

float_t namino_arancio::readNTC() {
    return readAnalogIn(5);
}

bool namino_arancio::writeAnalogInConf(int channel, uint8_t conf) {
    if (channel >= 1 && channel <=8) {
        reg[WR_ANALOG_IN_CH01_CONF + channel - 1] = conf;
        return true;
    }
    return false;
}

void namino_arancio::writeAnalogOut(float_t value) {
    switch(reg[WR_ANALOG_OUT_CH01_CONF]) {
        case OUT_CH01_DISABLED:
            reg[WR_ANALOG_OUT] = 0;
        break;
        case OUT_CH01_CURRENT:
            reg[WR_ANALOG_OUT] = value * 100; // place here um conversion
        break;
        case OUT_CH01_VOLTAGE:
            reg[WR_ANALOG_OUT] = value * 100; // place here um conversion
        break;
        default:
        break;
    }
}

void namino_arancio::writeAnalogOutConf(ANALOG_OUT_CH01_CONF_VALUES conf) {
    reg[WR_ANALOG_OUT_CH01_CONF] = conf;
}

void namino_arancio::writeRegister(uint8_t address, uint16_t value) {
    reg[address] = value;
}

bool namino_arancio::writeDigOut(int channel, bool value) {
    if (channel >= 1 && channel <= 8) {
      channel--;
      if (value) {
        reg[WR_DIG_OUT_1_16] |= 1UL << channel;
      } else {
        reg[WR_DIG_OUT_1_16] &= ~(1UL << channel);
      }
      return true;
    }
    return false;
}

void namino_arancio::writeRele(bool value) {
    reg[WR_RELE] = value;
}

void namino_arancio::setupStep(uint16_t frequency, uint32_t stepsNumber) {
  writeStepFreq(frequency);
  writeStepNumber(stepsNumber);
}

void namino_arancio::writeStepNumber(uint32_t value) {
    reg[WR_STEP_NUMBER_L] = value & 0x0000FFFF;
    reg[WR_STEP_NUMBER_H] = (value & 0xFFFF0000) >> 16;
}

void namino_arancio::writeStepFreq(uint16_t value) {
    reg[WR_STEP_FREQ] = value;
}

void namino_arancio::startStepper() {
  reg[WR_STEP_COMMAND] = 0x10;
}

void namino_arancio::stopStepper() {
  reg[WR_STEP_COMMAND] = 0x20;
}

void namino_arancio::readAllRegister() {
    reg[WR_WRITE_REQUEST] = 0;    // signal board firmware write direction indicator

    vspi->beginTransaction(spiSettings);
    digitalWrite(_ss, LOW);

    uint16_t *rp = reg;
    for (uint16_t i = 0; i <= WR_LAST_INDEX; i++) {
        *rp = vspi->transfer(0);
        *rp = (*rp << 8) | vspi->transfer(0);
        rp++;
    }

    digitalWrite(_ss, HIGH);
    vspi->endTransaction();
}

void namino_arancio::resetSignalMicroprocessor() {
  pinMode(NM_RESET, OUTPUT);
  digitalWrite(NM_RESET, !HIGH);
  delay(200);
  digitalWrite(NM_RESET, !LOW);
  delay(200);
  digitalWrite(NM_RESET, !HIGH);
}

void namino_arancio::writeAllRegister() {
    delay(10);
    reg[WR_WRITE_REQUEST] = 1;    // signal board firmware write direction indicator
    reg[WR_COUNTER]++;

    vspi->beginTransaction(spiSettings);
    digitalWrite(_ss, LOW);

    uint16_t *wp = reg;
    for (uint16_t i = 0; i <= WR_LAST_INDEX; i++) {
        vspi->transfer((*wp & 0xFF00) >> 8);
        vspi->transfer(*wp & 0xFF);
        wp++;
    }

    digitalWrite(_ss, HIGH);
    vspi->endTransaction();
}

void namino_arancio::transferAllRegister() {
    reg[WR_WRITE_REQUEST] = 1;    // signal board firmware write direction indicator
    reg[WR_COUNTER]++;

    vspi->beginTransaction(spiSettings);
    digitalWrite(_ss, LOW);

    uint16_t *wp = reg;
    for (uint16_t i = 0; i <= WR_LAST_INDEX; i++) {
        // DEBUG: Serial.printf("REG W %02d: 0x%04x ", i, *wp);
        uint16_t h = vspi->transfer((*wp & 0xFF00) >> 8);
        uint16_t l = vspi->transfer(*wp & 0xFF);
        // DEBUG: Serial.printf("H 0x%02x L 0x%02x ", h, l);
        *wp = (h << 8) | l;
        // DEBUG: Serial.printf("REG R %02d: 0x%04x ", i, *wp);
        // DEBUG: Serial.println();
        wp++;
    }

    digitalWrite(_ss, HIGH);
    vspi->endTransaction();
}

void namino_arancio::resetWRegister() {
    for (uint8_t i = WR_FIRST_INDEX; i <= WR_LAST_INDEX; i++) {
        reg[i] = 0;
    }
}

void namino_arancio::saveRegister(uint8_t address, uint16_t value) {
    reg[address] = value;
}

uint16_t namino_arancio::loadRegister(uint8_t address) {
    return reg[address];
}

void namino_arancio::showRegister() {
    uint16_t *rp = reg;
    for (uint8_t i = 0; i <= WR_LAST_INDEX; i += 4) {
            Serial.printf("REG %02d: 0x%04x REG %02d: 0x%04x REG %02d: 0x%04x REG %02d: 0x%04x\n",
                i + 0, *(rp + 0), 
                i + 1, *(rp + 1), 
                i + 2, *(rp + 2), 
                i + 3, *(rp + 3));
            rp += 4;
    }
    Serial.println("-----------------------------------------------------------");

    // uint16_t *wp = reg;
    // for (uint8_t i = WR_FIRST_INDEX; i <= WR_LAST_INDEX; i += 4) {
    //         Serial.printf("REG %02d: 0x%04x REG %02d: 0x%04x REG %02d: 0x%04x REG %02d: 0x%04x\n",
    //             i + 0, *(wp + 0), 
    //             i + 1, *(wp + 1), 
    //             i + 2, *(wp + 2), 
    //             i + 3, *(wp + 3));
    //         wp += 4;
    // }
    // Serial.println("W----------------------------------------------------------");
}

void namino_arancio::printValues() {
    static const char *separator = "----------------------------------|-------------------------|------------------------|-----------------------|--------------";

    for (uint8_t i = RO_ANALOG_IN_CH01; i <= RO_ANALOG_IN_CH05; i += 4) {
            Serial.printf("ANALOG IN CH %2d: %f CONF %2d | CH %2d: %f CONF %2d | CH %2d: %f CONF %d | CH %2d: %f CONF %d\n",
                i + 0 - RO_ANALOG_IN_CH01 + 1, readAnalogIn(i + 0), reg[WR_ANALOG_IN_CH01_CONF + i + 0],
                i + 1 - RO_ANALOG_IN_CH01 + 1, readAnalogIn(i + 1), reg[WR_ANALOG_IN_CH01_CONF + i + 1],
                i + 2 - RO_ANALOG_IN_CH01 + 1, readAnalogIn(i + 2), reg[WR_ANALOG_IN_CH01_CONF + i + 2],
                i + 3 - RO_ANALOG_IN_CH01 + 1, readAnalogIn(i + 3), reg[WR_ANALOG_IN_CH01_CONF + i + 3]);
    }
    Serial.println(separator);
    for (uint8_t i = 1; i <= 5; i += 4) {
            Serial.printf("DIGITAL IN CH %2d: %d               | CH %2d: %d                | CH %2d: %d               | CH %2d: %d\n",
                i + 0, readDigIn(i + 0),
                i + 1, readDigIn(i + 1),
                i + 2, readDigIn(i + 2),
                i + 3, readDigIn(i + 3));
    }
    Serial.println(separator);
    for (uint8_t i = 1; i <= 4; i += 4) {
            Serial.printf("DIGITAL OUT CH %2d: %d              | CH %2d: %d                | CH %2d: %d               | CH %2d: %d\n",
                i + 0, readDigOut(i + 0),
                i + 1, readDigOut(i + 1),
                i + 2, readDigOut(i + 2),
                i + 3, readDigOut(i + 3));
    }
    Serial.println(separator);
    Serial.printf("ENCODER %9d                 | ENC.SWITCH %1d \n", readEncoder(), readEncoderSwitch());
    Serial.println(separator);
}


src/Namino_Industrial_Boards.h
/*
MIT License

Copyright (c) 2023 Namino Team, version: 1.0.18 @ 2023-07-07

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

#include <namino_rosso.h>
#include <namino_arancio.h>


src/namino_arancio.h
/*
MIT License

Copyright (c) 2023 Namino Team, version: 1.0.18 @ 2023-07-07

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

#ifndef NAMINO_ARANCIO_H
#define NAMINO_ARANCIO_H

#include <Arduino.h>
#include <SPI.h>

/* INDUSTRIAL SIDE memory map register */
typedef enum {
  // Read only registers
  RO_NAMINO_ID            = 0,
  RO_LIFE_TIME_L          = 1,
  RO_LIFE_TIME_H          = 2,
  RO_RESERVED_3           = 3,
  RO_ANALOG_IN_CH01       = 4,
  RO_ANALOG_IN_CH02       = 5,
  RO_ANALOG_IN_CH03       = 6,
  RO_ANALOG_IN_CH04       = 7,
  RO_ANALOG_IN_CH05       = 8,
  RO_RESERVED_9           = 9,
  RO_RESERVED_10          = 10,
  RO_RESERVED_11          = 11,
  RO_RESERVED_12          = 12,
  RO_RESERVED_13          = 13,
  RO_DIG_IN_1_16          = 14,
  RO_ENCODER_COUNTER_L    = 15,
  RO_ENCODER_COUNTER_H    = 16,
  RO_ENCODER_SWITCH       = 17,
  RO_ANALOG_FRONT_END_ID  = 20,
  RO_RESERVED_21          = 21,
  RO_RESERVED_22          = 22,
  RO_RESERVED_23          = 23,
  RO_RESERVED_24          = 24,
  RO_RESERVED_25          = 25,
  RO_RESERVED_26          = 26,
  RO_RESERVED_27          = 27,
  RO_RESERVED_28          = 28,
  RO_RESERVED_29          = 29,
  RO_RESERVED_30          = 30,
  RO_LAST_INDEX           = 31,
  // Write registers
  WR_FIRST_INDEX          = 32,
  WR_WRITE_REQUEST        = WR_FIRST_INDEX + 0,   // 32
  WR_ENCODER_SET_L        = WR_FIRST_INDEX + 1,   // 33
  WR_ENCODER_SET_H        = WR_FIRST_INDEX + 2,   // 34
  WR_DIG_OUT_1_16         = WR_FIRST_INDEX + 3,   // 35
  WR_RELE                 = WR_FIRST_INDEX + 4,   // 36
  WR_STEP_NUMBER_L        = WR_FIRST_INDEX + 5,   // 37
  WR_STEP_NUMBER_H        = WR_FIRST_INDEX + 6,   // 38
  WR_STEP_FREQ            = WR_FIRST_INDEX + 7,   // 39
  WR_STEP_RESERVED        = WR_FIRST_INDEX + 8,   // 40
  WR_STEP_COMMAND         = WR_FIRST_INDEX + 9,   // 41
  WR_ANALOG_OUT           = WR_FIRST_INDEX + 10,  // 42
  WR_ANALOG_IN_CH01_CONF  = WR_FIRST_INDEX + 11,  // 43
  WR_ANALOG_IN_CH02_CONF  = WR_FIRST_INDEX + 12,  // 44
  WR_ANALOG_IN_CH03_CONF  = WR_FIRST_INDEX + 13,  // 45
  WR_ANALOG_IN_CH04_CONF  = WR_FIRST_INDEX + 14,  // 46
  WR_RESERVED_47          = WR_FIRST_INDEX + 15,  // 47
  WR_RESERVED_48          = WR_FIRST_INDEX + 16,  // 48
  WR_RESERVED_49          = WR_FIRST_INDEX + 17,  // 49
  WR_RESERVED_50          = WR_FIRST_INDEX + 18,  // 50
  WR_ANALOG_OUT_CH01_CONF = WR_FIRST_INDEX + 19,  // 51
  WR_COUNTER              = WR_FIRST_INDEX + 20,  // 52
  WR_CONFIG_REGISTER      = WR_FIRST_INDEX + 21,  // 53
  WR_RESERVED_54          = 54,
  WR_RESERVED_55          = 55,
  WR_RESERVED_56          = 56,
  WR_RESERVED_57          = 57,
  WR_RESERVED_58          = 58,
  WR_RESERVED_59          = 59,
  WR_RESERVED_60          = 60,
  WR_RESERVED_61          = 61,
  WR_RESERVED_62          = 62,
  WR_LAST_INDEX           = 63,
} NAMINO_ARANCIO_REGISTERS;

typedef enum {
  CH01_DISABLED    = 0,
  CH01_CURRENT     = 1,
  CH01_VOLTAGE     = 2,
} ANALOG_IN_CH01_CONF_VALUES;

typedef enum {
  CH02_DISABLED    = 0,
  CH02_CURRENT     = 1,
  CH02_VOLTAGE     = 2,
} ANALOG_IN_CH02_CONF_VALUES;

typedef enum {
  CH03_DISABLED    = 0,
  CH03_CURRENT     = 1,
  CH03_VOLTAGE     = 2,
} ANALOG_IN_CH03_CONF_VALUES;

typedef enum {
  CH04_DISABLED    = 0,
  CH04_CURRENT     = 1,
  CH04_VOLTAGE     = 2,
} ANALOG_IN_CH04_CONF_VALUES;

// jumper J6 (side A) configuration request
typedef enum {
  OUT_CH01_DISABLED    = 0,
  OUT_CH01_CURRENT     = 1,
  OUT_CH01_VOLTAGE     = 2,
} ANALOG_OUT_CH01_CONF_VALUES;

class namino_arancio {
public:
    namino_arancio();

    /**
     * open SPI interface.
     */
    void begin(uint32_t clock=800000U);

    /**
     * open SPI interface.
     */
    void begin(uint32_t clock=800000U, uint8_t miso=MISO, uint8_t mosi=MOSI, uint8_t sck=SCK, uint8_t ss=SS);

    /**
     * close SPI interface.
     */
    void end();

    /**
     * read namino arancio signal board, um converted
     * @param channel channel number, unit of measure as configuration
     */
    float_t readAnalogIn(int channel);

    /**
     * read namino arancio life time
     */
    uint32_t readLifeTime();

    /**
     * firmware version
     * @return signal board firmware version
     */
    uint16_t fwVersion();

    /**
     * board code identifier
     * @return identifier values 0x1338=rosso, 0x2904=arancio, ...
     */
    uint16_t boardType();

    /**
     * analog section is ready for configuration ?
     * @return true = ok, ready
     */
    bool isReady();

    /**
     * read namino arancio encoder
     */
    int32_t readEncoder();

    /**
     * read namino arancio encoder switch
     */
    bool readEncoderSwitch();

    /**
     * reset encoder counter
     */
    void resetEncoderCounter();

    /**
     * set encoder counter
     * @param value encoder counter
     */
    void setEncoderCounter(int32_t value);

    /**
     * read namino arancio digital input
     * @param channel channel number
     */
    bool readDigIn(int channel);

    /**
     * read namino arancio digital out feedback
     * @param channel channel number
     */
    bool readDigOut(int channel);

    /**
     * read namino arancio NTC in channel 5
     */
    float_t readNTC();

    /**
     * read all namino arancio signal board registers.
     */
    void readAllRegister();

    /**
     * full duplex read/write all namino arancio signal board registers.
     */
    void transferAllRegister();

    /**
     * write namino arancio analog in board configuration
     * @param channel channel number
     * @param conf conf values
     * @return true = success
     */
    bool writeAnalogInConf(int channel, uint8_t conf);


    /**
     * write namino arancio analog out value, voltage
     * @param value value 0..10 V
     */
    void writeAnalogOut(float_t value);  

    /**
     * write namino arancio analog out board configuration, jumper J6 (side A) configuration request
     * @param conf conf values
     */
    void writeAnalogOutConf(ANALOG_OUT_CH01_CONF_VALUES conf);

    /**
     * write namino arancio signal board registers.
     * @param address register number
     * @param value register value
     */
    void writeRegister(uint8_t address, uint16_t value);

    /**
     * reset write registers aread
     */
    void resetWRegister();

    /**
     * write namino arancio digital out
     * @param channel channel number
     * @param value out value
     * @return true = success
     */
    bool writeDigOut(int channel, bool value);

    /**
     * write namino arancio set rele status
     * @param value out value
     */
    void writeRele(bool value);

    /**
     * write namino arancio set PTO, pulse train output
     * @param frequency frequency of the pulse train, values ​​expressed in Hz admitted from 1 to 65535
     * @param stepsNumber number of steps that the PTO must carry out. If set to 0, the PTO does not finish counting until the stop command is received
     */
    void setupStep(uint16_t frequency, uint32_t stepsNumber);

    /**
     * write namino arancio set step number
     * @param value out value
     */
    void writeStepNumber(uint32_t value);

    /**
     * write namino arancio set step freq
     * @param value out value
     */
    void writeStepFreq(uint16_t value);

    /**
     * namino arancio start PTO
    // da note sul campo, comando:
    //     0x10 start -> esegue il numero di step indicati nel registro 36
    //     0x20 interrompe il treno in esecuzione, riarma il PTO per una nuova partenza
    // se si lancia un treno di impulsi, durante l'esecuzione il bit 8 (run) del registro 3 (naminoStatus) va a 1.
    // quando il treno è terminato, il registro 3 (naminoStatus) si modifica nel seguente modo: bit 8 (run) -> 0; bit 9 (done) -> 1
    // per riarmare il treno è necessario inviare il comando 0x20 (stop) e quindi il comando 0x10 (start)
    // Se frequenza = 0 la linea è utilizzata come uscita digitale out 
     */
    void startStepper();

    /**
     * namino arancio stop PTO
     * interrupts the running train, rearms the PTO for a new departure
     */
    void stopStepper();

    /**
     * write all namino arancio signal board registers.
     */
    void writeAllRegister();

    /**
     * register show, unformatted
     */
    void showRegister();

    /**
     * register show, formatted
     */
    void printValues();

    /*
     * reset namino microprosessor side
     */
    void resetSignalMicroprocessor();

    /**
     * save in register cache.
     * @param address register number
     * @param value register value
     */
    void saveRegister(uint8_t address, uint16_t value);

    /**
     * load from register cache.
     * @param address register number
     * @result cache register value
     */
    uint16_t loadRegister(uint8_t address);

private:
    uint8_t _miso;
    uint8_t _mosi;
    uint8_t _sck;
    uint8_t _ss;
    SPISettings spiSettings;
    SPIClass *vspi = NULL;

    uint16_t reg[WR_LAST_INDEX + 1];    // in / out register array
};

#endif  /* NAMINO_ARANCIO_H */
